src/app/mod.rs

```diff
@@ -457,4 +661,107 @@ impl App {
             Ok(())
         }
     }
+
+    /// # Errors
+    ///
+    /// Will return `Err` if the provided proof is invalid.
+    #[instrument(level = "debug", skip(self))]
+    pub async fn verify_semaphore_proof_v2(
+        &self,
+        root: Field,
+        signal_hash: Field,
+        nullifier_hash: Field,
+        external_nullifier_hash: Field,
+        proof: Proof,
+        max_root_age_seconds: Option<i64>,
+    ) -> Result<bool, VerifySemaphoreProofV2Error> {
+        let Some(root_state) = self.database.get_root_state(&root).await? else {
+            return Err(VerifySemaphoreProofV2Error::InvalidRoot);
+        };
+
+        if let Some(max_root_age_seconds) = max_root_age_seconds {
+            let max_root_age = Duration::seconds(max_root_age_seconds);
+            if !self.validate_root_age_v2(max_root_age, &root_state)? {
+                return Err(VerifySemaphoreProofV2Error::RootTooOld);
+            }
+        }
+
+        let proof = if Self::is_proof_padded(&proof) {
+            let proof_flat = proof.flatten();
+            let compressed_flat = [proof_flat[0], proof_flat[1], proof_flat[2], proof_flat[3]];
+            let compressed = CompressedProof::from_flat(compressed_flat);
+            semaphore_rs::protocol::compression::decompress_proof(compressed)
+                .ok_or_else(|| VerifySemaphoreProofV2Error::DecompressingProofError)?
+        } else {
+            proof
+        };
+
+        verify_proof(
+            root,
+            nullifier_hash,
+            signal_hash,
+            external_nullifier_hash,
+            &proof,
+            self.config.tree.tree_depth,
+        )
+        .map_err(|_err| VerifySemaphoreProofV2Error::ProverError)
+    }
+
+    /// We consider root age valid in two cases:
+    /// * root is current root of latest, processed or mined tree
+    /// * root age (depending on state) is not greater than max_root_age
+    fn validate_root_age_v2(
+        &self,
+        max_root_age: Duration,
+        root_state: &RootItem,
+    ) -> anyhow::Result<bool> {
+        let tree_state = self.tree_state()?;
+        let latest_root = tree_state.get_latest_tree().get_root();
+        let processed_root = tree_state.get_processed_tree().get_root();
+        let mined_root = tree_state.get_mined_tree().get_root();
+
+        info!("Validating age max_root_age: {max_root_age:?}");
```

Comment:
```suggestion
        info!(?max_root_age, "Validating age max_root_age");
```

or something like this, I forgot the exact syntax

Reply:
Do we even need this log as we are removing all the others? I'm gonna remove it.